File: packages/uns-core/src/tools/auth/auth-client.ts
import { ConfigFile } from "../../config-file";
import { SecureStoreFactory, ISecureStore } from "./secure-store";
import jwt from "jsonwebtoken";
import readline from "readline";

type LoginResponse = {
  accessToken: string;
};

/**
 * AuthClient handles acquiring and refreshing JWT access tokens
 * using the configured REST base URL.
 */
export class AuthClient {
  private readonly restBase: string;
  private readonly namespace: string;
  private store!: ISecureStore;

  private constructor(restBase: string) {
    this.restBase = restBase.replace(/\/$/, "");
    // namespace store by rest base to allow multiple environments
    this.namespace = `uns-auth:${this.restBase}`;
  }

  static async create(): Promise<AuthClient> {
    const cfg = await ConfigFile.loadConfig();
    const restBase: string = cfg?.uns?.rest;
    if (!restBase) throw new Error("config.uns.rest is not set");
    const client = new AuthClient(restBase);
    client.store = await SecureStoreFactory.create(client.namespace);
    return client;
  }

  async getAccessToken(): Promise<string> {
    const accessToken = await this.store.get("accessToken");
    const refreshToken = await this.store.get("refreshToken");

    if (accessToken && !AuthClient.isExpired(accessToken)) {
      return accessToken;
    }

    // Try refresh if we have refresh token
    if (refreshToken) {
      try {
        const refreshed = await this.refresh(refreshToken);
        await this.persistTokens(refreshed.accessToken, refreshed.refreshToken);
        return refreshed.accessToken;
      } catch {
        // ignore, fallback to login
      }
    }

    // Interactive login
    const { email, password } = await AuthClient.promptCredentials();
    const loggedIn = await this.login(email, password);
    await this.persistTokens(loggedIn.accessToken, loggedIn.refreshToken);
    return loggedIn.accessToken;
  }

  private static isExpired(token: string, skewSeconds = 30): boolean {
    try {
      const decoded: any = jwt.decode(token);
      if (!decoded || typeof decoded !== "object" || !decoded.exp) return true;
      const now = Math.floor(Date.now() / 1000);
      return decoded.exp <= now + skewSeconds;
    } catch {
      return true;
    }
  }

  private static endpoint(base: string, tail: string): string {
    // base ends without trailing slash; tail must not start with slash
    const b = base.replace(/\/$/, "");
    const t = tail.replace(/^\//, "");
    return `${b}/${t}`;
  }

  private static async fetchWithTimeout(url: string, init: RequestInit = {}, timeoutMs = 10_000): Promise<Response> {
    const controller = new AbortController();
    const id = setTimeout(() => controller.abort(), timeoutMs);
    try {
      const resp = await fetch(url, { ...init, signal: controller.signal });
      return resp;
    } finally {
      clearTimeout(id);
    }
  }

  private static extractRefreshCookie(resp: Response): string | null {
    const anyHeaders: any = resp.headers as any;
    const getSetCookie = typeof anyHeaders.getSetCookie === "function" ? anyHeaders.getSetCookie.bind(anyHeaders) : null;
    const candidates: string[] = getSetCookie ? getSetCookie() : (resp.headers.get("set-cookie") ? [resp.headers.get("set-cookie") as string] : []);
    for (const header of candidates) {
      const match = header.match(/(?:^|;\s*)RefreshToken=([^;]+)/i);
      if (match) return match[1];
    }
    return null;
  }

  private async login(email: string, password: string): Promise<{ accessToken: string; refreshToken: string; }> {
    const url = AuthClient.endpoint(this.restBase, "auth/login");
    const resp = await AuthClient.fetchWithTimeout(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ email, password }),
    }, 12_000);

    if (!resp.ok) {
      throw new Error(`Login failed: ${resp.status} ${resp.statusText}`);
    }
    const data = (await resp.json()) as LoginResponse;
    const refreshToken = AuthClient.extractRefreshCookie(resp);
    if (!data?.accessToken || !refreshToken) {
      throw new Error("Login response missing tokens");
    }
    return { accessToken: data.accessToken, refreshToken };
  }

  private async refresh(refreshToken: string): Promise<{ accessToken: string; refreshToken: string; }> {
    const url = AuthClient.endpoint(this.restBase, "auth/refresh");
    const resp = await AuthClient.fetchWithTimeout(url, {
      method: "POST",
      headers: {
        // server expects cookie RefreshToken
        "Cookie": `RefreshToken=${refreshToken}`,
      },
    }, 8_000);

    if (!resp.ok) {
      throw new Error(`Refresh failed: ${resp.status} ${resp.statusText}`);
    }
    const data = (await resp.json()) as LoginResponse;
    const newRefreshToken = AuthClient.extractRefreshCookie(resp) || refreshToken;
    if (!data?.accessToken) throw new Error("Refresh response missing accessToken");
    return { accessToken: data.accessToken, refreshToken: newRefreshToken };
  }

  private async persistTokens(accessToken: string, refreshToken: string): Promise<void> {
    await this.store.set("accessToken", accessToken);
    await this.store.set("refreshToken", refreshToken);
  }

  static async promptCredentials(): Promise<{ email: string; password: string }> {
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout, terminal: true });

    const ask = (q: string) => new Promise<string>((resolve) => rl.question(q, (ans) => resolve(ans.trim())));

    const askMasked = (q: string) => new Promise<string>((resolve) => {
      const anyRl = rl as any;
      const out: any = anyRl.output || process.stdout;
      const origWrite = (anyRl._writeToOutput as ((str: string) => void))?.bind(rl) || ((s: string) => out.write(s));
      anyRl.stdoutMuted = true;
      anyRl._writeToOutput = function (stringToWrite: string) {
        // Keep prompt text intact; mask user input
        if (anyRl.stdoutMuted) {
          if (stringToWrite.startsWith(q)) {
            out.write(stringToWrite);
          } else if (stringToWrite.endsWith("\n")) {
            out.write("\n");
          } else {
            out.write("*");
          }
        } else {
          origWrite(stringToWrite);
        }
      };
      rl.question(q, (value) => {
        anyRl.stdoutMuted = false;
        anyRl._writeToOutput = origWrite;
        out.write("\n");
        resolve(value.trim());
      });
    });

    try {
      const email = await ask("Email: ");
      const password = await askMasked("Password: ");
      rl.close();
      return { email, password };
    } catch (e) {
      rl.close();
      throw e;
    }
  }
}

File: packages/uns-py/src/uns_kit/core/auth_client.py
from __future__ import annotations

import base64
import getpass
import json
import time
import urllib.error
import urllib.request
from http.cookiejar import CookieJar
from pathlib import Path
from typing import Any, Optional

from .config_file import ConfigFile
from .secure_store import SecureStore, SecureStoreFactory


class AuthClient:
    def __init__(
        self,
        rest_base: str,
        *,
        email: Optional[str] = None,
        password: Optional[str] = None,
        store: Optional[SecureStore] = None,
        timeout: float = 10.0,
    ) -> None:
        self.rest_base = rest_base.rstrip("/")
        self.namespace = f"uns-auth:{self.rest_base}"
        self.email = email
        self.password = password
        self.timeout = timeout
        self.access_token: Optional[str] = None
        self.refresh_token: Optional[str] = None
        self.store = store or SecureStoreFactory.create(self.namespace)
        self._cookies = CookieJar()
        self._opener = urllib.request.build_opener(urllib.request.HTTPCookieProcessor(self._cookies))

    @classmethod
    def create(
        cls,
        config_path: Optional[Path | str] = None,
        *,
        timeout: float = 10.0,
    ) -> "AuthClient":
        cfg = ConfigFile.load_config(config_path)
        uns = cfg.get("uns") if isinstance(cfg.get("uns"), dict) else {}
        rest_base = uns.get("rest")
        if not isinstance(rest_base, str) or not rest_base:
            raise ValueError("config.uns.rest is not set")
        email = uns.get("email") if isinstance(uns.get("email"), str) else None
        password = uns.get("password") if isinstance(uns.get("password"), str) else None
        return cls(rest_base, email=email, password=password, timeout=timeout)

    def get_access_token(self) -> str:
        if self.access_token is None:
            self.access_token = self.store.get("accessToken")
        if self.refresh_token is None:
            self.refresh_token = self.store.get("refreshToken")

        if self.access_token and not self.is_expired(self.access_token):
            return self.access_token

        if self.refresh_token:
            try:
                return self.refresh_access_token()
            except Exception:
                pass

        if self.email and self.password:
            tokens = self.login(self.email, self.password)
            return tokens["accessToken"]

        email, password = self.prompt_credentials()
        tokens = self.login(email, password)
        return tokens["accessToken"]

    def login(self, email: str, password: str) -> dict[str, Any]:
        response = self._request_json(
            "POST",
            self._endpoint("auth/login"),
            body={"email": email, "password": password},
        )
        access_token = response.get("accessToken")
        refresh_token = self._read_refresh_token_cookie()
        if not isinstance(access_token, str) or not refresh_token:
            raise RuntimeError("Login response missing tokens")
        self.email = email
        self.password = password
        self.access_token = access_token
        self.refresh_token = refresh_token
        self._persist_tokens(access_token, refresh_token)
        return {
            **response,
            "accessToken": access_token,
            "refreshToken": refresh_token,
        }

    def refresh_access_token(self) -> str:
        if not self.refresh_token:
            raise RuntimeError("No refresh token available")
        response = self._request_json(
            "POST",
            self._endpoint("auth/refresh"),
            headers={"Cookie": f"RefreshToken={self.refresh_token}"},
        )
        access_token = response.get("accessToken")
        if not isinstance(access_token, str):
            raise RuntimeError("Refresh response missing accessToken")
        self.access_token = access_token
        self.refresh_token = self._read_refresh_token_cookie() or self.refresh_token
        if self.refresh_token:
            self._persist_tokens(self.access_token, self.refresh_token)
        return access_token

    def clear_tokens(self) -> None:
        self.access_token = None
        self.refresh_token = None
        self.store.delete("accessToken")
        self.store.delete("refreshToken")

    @staticmethod
    def is_expired(token: str, skew_seconds: int = 30) -> bool:
        try:
            segments = token.split(".")
            if len(segments) < 2:
                return True
            payload_segment = segments[1]
            padding = "=" * (-len(payload_segment) % 4)
            decoded = base64.urlsafe_b64decode(f"{payload_segment}{padding}".encode("ascii"))
            payload = json.loads(decoded.decode("utf-8"))
            exp = payload.get("exp")
            if not isinstance(exp, (int, float)):
                return True
            return float(exp) <= time.time() + skew_seconds
        except Exception:
            return True

    @staticmethod
    def prompt_credentials() -> tuple[str, str]:
        email = input("Email: ").strip()
        password = getpass.getpass("Password: ").strip()
        return email, password

    def _endpoint(self, tail: str) -> str:
        return f"{self.rest_base}/{tail.lstrip('/')}"

    def _request_json(
        self,
        method: str,
        url: str,
        *,
        body: Optional[dict[str, Any]] = None,
        headers: Optional[dict[str, str]] = None,
    ) -> dict[str, Any]:
        data = None if body is None else json.dumps(body).encode("utf-8")
        request = urllib.request.Request(
            url,
            data=data,
            method=method,
            headers={
                "Accept": "application/json",
                **({"Content-Type": "application/json"} if data is not None else {}),
                **(headers or {}),
            },
        )
        try:
            with self._opener.open(request, timeout=self.timeout) as response:
                decoded = response.read().decode("utf-8")
        except urllib.error.HTTPError as exc:
            raise RuntimeError(f"{method} {url} failed: {exc.code} {exc.reason}") from exc
        except urllib.error.URLError as exc:
            raise RuntimeError(f"{method} {url} failed: {exc.reason}") from exc

        if not decoded:
            return {}
        parsed = json.loads(decoded)
        if not isinstance(parsed, dict):
            raise RuntimeError("Auth response must be a JSON object")
        return parsed

    def _read_refresh_token_cookie(self) -> Optional[str]:
        for cookie in self._cookies:
            if cookie.name in {"RefreshToken", "rt"}:
                return cookie.value
        return None

    def _persist_tokens(self, access_token: str, refresh_token: str) -> None:
        self.store.set("accessToken", access_token)
        self.store.set("refreshToken", refresh_token)
